All_Relationships/4.data-disc-model.py

```diff
@@ -0,0 +1,234 @@
+
+
+# In[ ]:
+
+
+print "Total Data Shape:"
+print F_train.shape
+print F_dev.shape
+print
+
+
+# # Train Sparse Logistic Regression Disc Model
+
+# Here we train an SLR. To find the optimal hyperparameter settings this code uses a [random search](http://scikit-learn.org/stable/modules/grid_search.html) instead of iterating over all possible combinations of parameters. After the final model has been found, it is saved in the checkpoints folder to be loaded in the [next notebook](5.data-analysis.ipynb). Furthermore, the weights for the final model are output into a text file to be analyzed as well.
+
+# In[ ]:
+
+
+get_ipython().magic(u'time train_marginals = load_marginals(session, split=0)')
+
+
+# In[ ]:
+
+
+# Searching over learning rate
+param_ranges = {
+    'lr' : [1e-2, 1e-3, 1e-4, 1e-5, 1e-6],
+    'l1_penalty' : [1e-2, 1e-3, 1e-4, 1e-5, 1e-6],
+    'l2_penalty' : [1e-2, 1e-3, 1e-4, 1e-5, 1e-6]
+}
+model_hyperparams = {
+    'n_epochs' : 50,
+    'rebalance' : 0.5,
+    'print_freq' : 25
+}
+searcher = RandomSearch(SparseLogisticRegression, param_ranges, F_train,
+                        Y_train=train_marginals, n=5, model_hyperparams=model_hyperparams)
+
+
+# In[ ]:
+
+
+get_ipython().run_cell_magic(u'time', u'', u'np.random.seed(100)\ndisc_model, run_stats = searcher.fit(F_dev, L_dev, n_threads=4)')
+
+
+# In[ ]:
+
+
+w, b = disc_model.get_weights()
+
+
+# In[ ]:
+
+
+# Write the weights and features for further processing
+annot_select_query = FeatureKey.__table__.select().order_by(FeatureKey.id)
+with open("LR_model.csv", "w") as f:
+    fieldnames = ["Weight", "Feature"]
+    writer = csv.DictWriter(f, fieldnames=fieldnames)
+    writer.writeheader()
+    for weight, feature in tqdm.tqdm(zip(w, session.execute(annot_select_query))):
+        writer.writerow({"Weight": weight, "Feature":feature[1]})
+
+
+# ## Train a LSTM Disc Model
+
+# This block of code trains an LSTM. An LSTM is a special type of recurrent nerual network that retains a memory of past values over period of time. ([Further explaination here](http://colah.github.io/posts/2015-08-Understanding-LSTMs/)). The problem with the code below is that sqlalchemy runs into an out of memory error on my computer during the preprocessing step. As a consequence we have to resort loading this data onto University of Pennsylvania's Performance Computing Cluster. The data that gets preprocessed is exported to a text file and then get shipped towards the cluster.
+
+# In[ ]:
+
+
+get_ipython().magic(u'time train_marginals = load_marginals(session, split=0)')
+np.savetxt("pmacs/train_marginals", train_marginals)
+
+
+# In[ ]:
+
+
+get_ipython().run_cell_magic(u'time', u'', u"train_kwargs = {\n    'lr':         0.001,\n    'dim':        100,\n    'n_epochs':   10,\n    'dropout':    0.5,\n    'print_freq': 1,\n    'max_sentence_length': 1000,\n}\n\nlstm = reRNN(seed=100, n_threads=4)\n#lstm.train(train_cands, train_marginals[0:10], X_dev=dev_cands, Y_dev=L_dev[0:10], **train_kwargs)")
```

Comment:
`train_kwargs` is never used? The command that uses it is commented out... should this be commented out too?

Reply:
`train_kwargs` should be commented out since I don't use it here. The initialization of the LSTM object shouldn't be commented out since I need to use it to generate the training data for the cluster.